### ISSUE TITLE
[Bug]: DeepSeek-V2-Lite with online FP8 on an RTX 4090 (TRITON_MLA): NaN logits on every request once three or more sequences are decoded in one CUDA-graph batch

### ISSUE THREAD

**Issue by Yunzez** [vllm-project/vllm#57017]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
((vllm-27) ) gpu-a6-7:~$ python -m vllm.collect_env
Collecting environment information...
==============================
        System Info
==============================
OS                           : Red Hat Enterprise Linux release 8.10 (Ootpa) (x86_64)
GCC version                  : (GCC) 8.5.0 20210514 (Red Hat 8.5.0-28)
Clang version                : Could not collect
CMake version                : version 4.3.1
Libc version                 : glibc-2.28

==============================
       PyTorch Info
==============================
PyTorch version              : 2.10.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.12 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 20:16:04) [GCC 11.2.0] (64-bit runtime)
Python platform              : Linux-4.18.0-553.89.1.el8_10.x86_64-x86_64-with-glibc2.28

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   :
GPU models and configuration : GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version        : 590.44.01
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  1
Core(s) per socket:  48
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               143
Model name:          Intel(R) Xeon(R) Platinum 8468
Stepping:            8
CPU MHz:             3800.000
CPU max MHz:         3800.0000
CPU min MHz:         800.0000
BogoMIPS:            4200.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            2048K
L3 cache:            107520K
NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.5.2
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.22.1
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.4.2
[pip3] nvidia-cutlass-dsl-libs-base==4.4.2
[pip3] nvidia-ml-py==13.595.45
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.4.5
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0+cu128
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.25.0+cu128
[pip3] transformers==4.57.6
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.27.1
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  	GPU0	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NODE	0,2,4,6,8,10	0		N/A
NIC0	NODE	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_user

((vllm-27) ) gpu-a6-7:~$
```

</details>

### 🐛 Describe the bug

**Environment**

- vLLM 0.27.1 (pip wheel), torch 2.13.0+cu130, CUDA 13.0, driver 580.126.09, Linux 5.15.
- One RTX 4090 (sm_89). Attention: `TRITON_MLA` (the only MLA backend on sm_89) with the `FLASH_ATTN` MLA
  prefill backend; `TRITON Fp8 MoE`; `CutlassFP8ScaledMMLinearKernel`.
- Server: `vllm serve deepseek-ai/DeepSeek-V2-Lite-Chat --trust-remote-code --quantization fp8 --max-model-len 4096`
  (default execution mode: torch.compile + CUDA graphs; prefix caching on). The KV pool size and
  `--num-gpu-blocks-override` make no difference (checked both ways).

As soon as **three or more** sequences are decoded in the same batch, **every request in the batch returns NaN log-probs** (HTTP 400 "Out of range float values are not JSON compliant: nan" when `logprobs` is requested; without `logprobs` the text is garbage, e.g. `"should should should ..."`). The engine logs nothing: no warning, no error, just NaN in the logits.

| concurrent greedy requests (prompt 10-115 words, decode 16-640 tokens) | default (compile + CUDA graphs) | `--enforce-eager` |
|---|---|---|
| 1, or 2 (joining during the first's prefill or its decode) | fine | fine |
| 3 (three repeats), 4, 5, 8, 16, 24, 32 | NaN on every request (in some 4-request rounds one survives) | fine, 3 and 5 alike |

Preemption is not involved (the preemption counter stays at 0 up to 8 concurrent requests). Isolating the execution mode:

| server | K=2 | K=3 | K=5 |
|---|---|---|---|
| default (torch.compile + CUDA graphs) | fine | NaN x3 | NaN x5 |
| same, default KV pool size | fine | NaN x3 | NaN x5 |
| torch.compile on, `--compilation-config '{"cudagraph_mode":"NONE"}'` | fine | **fine** | **fine** |
| `--enforce-eager` | fine | fine | fine |
| default mode, `--max-num-seqs 2` | fine | fine (never three in a batch) | fine |

So the NaN comes from **CUDA-graph replay of a decode batch of three or more sequences** on the `TRITON_MLA` + online-FP8 path; inductor alone is fine and the block pool is irrelevant. `--enforce-eager` or `cudagraph_mode: NONE` fixes it; `--max-num-seqs 2` sidesteps it.

Online FP8 is the documented way to fit this model on a 24 GB card (the bf16 checkpoint is 32 GB), and the default configuration produces garbage for every request the moment the server has three in flight. 

### Reproducer

Standalone (only `httpx`), against a default server on one 4090: 1, 2, then 3 concurrent 16-token greedy
requests with `logprobs`. Expected: three HTTP 200 rounds. Observed: 200, 200, then three HTTP 400
(`... not JSON compliant: nan`); with `--enforce-eager`, three 200s.

    vllm serve deepseek-ai/DeepSeek-V2-Lite-Chat --trust-remote-code --quantization fp8 --max-model-len 4096 --port 8000
    python3 repro_dsv2lite_fp8_nan.py --base-url http://127.0.0.1:8000        # --k 3 (default), try --k 5

```python
# Standalone reproducer (no dependency on this repo): DeepSeek-V2-Lite-Chat with online FP8 on an
# RTX 4090 (sm89, TRITON_MLA) returns NaN log-probs / garbage text as soon as three requests are
# decoded in the same batch; two are fine; --enforce-eager or --compilation-config '{"cudagraph_mode":"NONE"}' is fine
# (torch.compile without CUDA graphs), so the fault is in CUDA-graph replay of a >=3-sequence decode batch.
#
#   vllm serve deepseek-ai/DeepSeek-V2-Lite-Chat --trust-remote-code --quantization fp8 --max-model-len 4096 --port 8000
#   python3 repro_dsv2lite_fp8_nan.py --base-url http://127.0.0.1:8000 [--k 3]
import argparse, math, random, threading
import httpx


def prompt(seed, n_words=80):
    rng = random.Random(seed)
    return " ".join(f"w{rng.randrange(10000)}" for _ in range(n_words))


def one(base, model, seed, n, out):
    r = httpx.post(base + "/v1/completions", json={"model": model, "prompt": prompt(seed), "max_tokens": n,
                                                    "temperature": 0.0, "logprobs": 1}, timeout=600)
    if r.status_code != 200:
        out[seed] = f"HTTP {r.status_code}: {r.text[:90]}"
        return
    ch = r.json()["choices"][0]
    lps = ch["logprobs"]["token_logprobs"]
    nan = sum(1 for x in lps if x is None or (isinstance(x, float) and math.isnan(x)))
    out[seed] = f"HTTP 200, nan_logprobs={nan}/{len(lps)}, text={ch['text'][:32]!r}"


ap = argparse.ArgumentParser()
ap.add_argument("--base-url", default="http://127.0.0.1:8000")
ap.add_argument("--model", default="deepseek-ai/DeepSeek-V2-Lite-Chat")
ap.add_argument("--k", type=int, default=3)
ap.add_argument("--max-tokens", type=int, default=16)
a = ap.parse_args()
for K in (1, 2, a.k):
    out, ths = {}, []
    for s in range(K):
        th = threading.Thread(target=one, args=(a.base_url, a.model, 100 * K + s, a.max_tokens, out))
        th.start(); ths.append(th)
    for th in ths:
        th.join()
    print(f"{K} concurrent request(s):")
    for s in sorted(out):
        print(f"   {out[s]}")
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by dundysm** [comment]
taking a look; planning to isolate the TRITON_MLA + online FP8 CUDA-graph decode path for batch size >= 3 (eager / cudagraph_mode NONE already look clean per the report)

**Comment by dundysm** [comment]
opened a fix pr: https://github.com/vllm-project/vllm/pull/57094

root cause: cudagraph decode pads inactive slots with seq_len=0; triton_decode_attention stage2 did acc/e_sum with e_sum=0 (nan). online fp8 moe then takes a per-tensor amax over the batch, so one pad nan poisons every request. the pr writes zeros when e_sum==0 and adds a kernel regression test (3 active + 1 zero-len pad).

verified on gpu (nvidia rtx 2000 ada, cuda 12.8, torch 2.8.0+cu128):

```
pytest tests/kernels/attention/test_triton_decode_attention.py -k zero_seq_len_pad_no_nan -v
# 3 passed, 115 deselected in 2.90s
```

all three `test_decode_attention_zero_seq_len_pad_no_nan[...]` cases passed (mla/gqa/mha pad shapes).
